Add Pots V2

dannysmith opened the pull request:
This is still a work in progress, but thought I'd open a PR now incase you have any comments - I'll clean up in the end, obviously.

I kinda feel like `deposit` and `withdraw` should be instance methods on `Pot`, rather than class methods, but that's not how the rest of the library works. What are your thoughts, @murraysum?

dannysmith committed:
Add PUT method to Client

dannysmith committed:
Switch to V2 pots

dannysmith committed:
Add generic API Error class

We'llw ant to raise an exception later on, and should use a specific class so it can be caught by user applications.

dannysmith committed:
Add withdraw, find and deposit methods to pots

Find mirrors the structure in Transaction#find, while deposit and withdraw are both instance methods on Pot, which return the updated instance.

dannysmith committed:
Remove redundant client variable

dannysmith committed:
Add tests for Pot::find, Pot#deposit and Pot#withdraw

dannysmith commented:
I've turned `deposit` and `withdraw` into instance methods on `Monzo::Pot`, because it seems to make much more sense.

These methods modify the `Pot` instance, but if you think that's a bad idea I'm open to changing them so they return a new instance instead. Alternatively, we could have both a bang! and non-bang! versions of the methods.

Let me know what you think, and thanks for making this 🙂

dannysmith committed:
Add new Pot methods to README

dannysmith committed:
Bugfix 🙃

dannysmith committed:
Require errors file. Duh

dannysmith committed:
Another bugfix

murraysum commented on lib/monzo/errors.rb:
```diff
@@ -0,0 +1,6 @@
+
+module Monzo
+  # Internal: Exception to raise on reciept of API error.
+  MonzoAPIError = Class.new(StandardError)
```
@dannysmith I think it would be worth drop the monzo part of the class name as it is in already in the Monzo namespace. I'd suggest renaming it to `ApiError`. Thoughts?

dannysmith replied:
Good shout.

murraysum commented on README.md:
```diff
@@ -63,6 +63,25 @@ A Pot is a place to keep some money separate from your main spending account.
 Monzo::Pot.all
+
+# Find a pot with the given pot id.
+Monzo::Pots.find(pot_id)
```
@dannysmith You have a spelling mistake here. It should be `Monzo::Pot.find(pot_id)`

murraysum commented on lib/monzo/pot.rb:
```diff
@@ -25,13 +27,89 @@ def initialize(params)
+      response = Monzo.client.get("/pots/#{pot_id}")
+      parsed_response = JSON.parse(response.body, :symbolize_names => true)
+
+      Monzo::Pot.new(parsed_response[:pot])
```
@dannysmith I don't believe the API returns JSON with a key of `pot`.

So instead of `Monzo::Pot.new(parsed_response[:pot])` this should be `Monzo::Pot.new(parsed_response)`.

dannysmith replied:
When I wrote this originally, the `/pots` endpoint was undocumented and returned an object wrapped in "pot" key. It doesn't any more (which is much more sensible).

murraysum commented on README.md:
```diff
@@ -63,6 +63,25 @@ A Pot is a place to keep some money separate from your main spending account.
+pot = Monzo::Pot.all.first # Get the first pot
+pot.balance #=> eg. 5000
+
+pot.deposit(100, account_id)
```
@dannysmith You have a spelling mistake here. It should be `deposit!` as it is a bang method.

murraysum commented on README.md:
```diff
@@ -63,6 +63,25 @@ A Pot is a place to keep some money separate from your main spending account.
+pot = Monzo::Pot.all.first
+pot.balance #=> eg. 5000
+
+pot.withdraw(100, account_id)
```
@dannysmith You have a spelling mistake here. It should be `withdraw!` as it is a bang method.

murraysum commented on lib/monzo/pot.rb:
```diff
@@ -25,13 +27,89 @@ def initialize(params)
+    #
+    # amount - The amount to deposit, in pennies.
+    # source_account_id - The account_id of the account to withdraw from.
+    # dedupe_id (optional) - A random string, to prevent duplicate deposits.
```
@dannysmith I am wondering what your thoughts on whether this should be optional. From reading the Monzo docs it would appear `dedupe_id` should be:

> A unique string used to de-duplicate deposits. Ensure this remains static between retries to ensure only one deposit is created.

In your implementation a new dedupe_id is created every time `deposit!` or `withdraw!` is called which saves the user having to think about generating a random id. However it does mean a user of the client could create accidental duplicate requests on retries. I wonder if it would be better to reveal this more to the user by providing a deduplication id generator and then making the `dedupe_id` a required field. 

```
dedupe_id = Monzo::Duplication.create_id

pot.withdraw(100, account_id, dedupe_id)

# A retry with the same dedupe id
pot.withdraw(100, account_id, dedupe_id)

```

What do you think?

dannysmith replied:
I thought about this a bunch and decided against requiring it. I reckon ease of use outweighs the need for dupe protection, particularly considering the type of things people are building on the Monzo API at the moment (ie. mainly toy apps for their own use). I'd hope that anyone building something that's retrying failed transactions would be sensible enough to include `dedupe_id`.

That said, we can't expect everyone to magically know it's an available argument so I'm about to add a section to the README by way of explaination.

Let me know if you disagree with this thinking and I'll make it a required argument.

Either way, I don't think we should provide a generator as part of this gem - the job of generating random strings seems a bit outside its remit, and Ruby's `SecureRandom` does the job easily enough.

murraysum replied:
@dannysmith Yeah everything you have said crossed my mind as well. Glad we are on the same page about the tradeoffs. 👍 

I think lets ship it as being an optional argument. If we get people complaining it is unexpected we can revisit it later.

Thanks for your thoughts and opinions. 🙌

murraysum reviewed (changes requested):
Hey @dannysmith 

Thanks for putting this together. Sorry it has taken me so long to get back to you, things have been a bit hectic around here.

I've noticed a few spelling mistakes, a bug and I have added some thoughts around the deduplication id. I'd love to get your feedback on those things. Apart from that this is looking pretty smashing to me.

Let me know your thoughts and hopefully over the next week we can put together a new release. Thanks again for your help 🙌 🎉

dannysmith committed:
Add note on deduplication to README

dannysmith committed:
Fix typos in README

dannysmith committed:
Rename Monzo::MonzoAPIError

dannysmith committed:
Fix typo in comment

dannysmith committed:
Fix bug in Pot::find

When I wrote this originally, the /pots endpoint was undocumented and returned an object wrapped in :pots. It doesn't any more, which is much more sensible.

dannysmith commented:
Hey @murraysum 👋 

> Thanks for putting this together. Sorry it has taken me so long to get back to you, things have been a bit hectic around here.

Same here!

I've pushed some fixes and dropped a few comments above. Let me know what you think about the `dedupe_id` thing.

murraysum commented:
@dannysmith Hey this is looking super duper to me. 🙌  Thanks for taking the time to address the feedback I left.. I'll get a release out as soon as possible.

murraysum commented:
@dannysmith That is the release out: https://rubygems.org/gems/monzo/versions/0.3.0

Thanks again for all your hard work. 🎉

dannysmith commented:
No probs.  Thanks for making the gem!
